taxcalc/parameters_base.py

```diff
@@ -294,26 +293,75 @@ def _update(self, year_mods):
             if name.endswith('_cpi'):
-                continue
-            if name in paramvals:
-                default_cpi = paramvals[name].get('cpi_inflated', False)
+                continue  # handle elsewhere in this method
+            if name in self._vals:
+                vals_indexed = self._vals[name].get('cpi_inflated', False)
             else:
-                default_cpi = False
-            cpi_inflated = year_mods[year].get(name + '_cpi', default_cpi)
+                msg = 'parameter name {} not in parameter values dictionary'
+                raise ValueError(msg.format(name))
+            name_plus_cpi = name + '_cpi'
+            if name_plus_cpi in year_mods[year].keys():
+                indexed = year_mods[year].get(name_plus_cpi)
+                self._vals[name]['cpi_inflated'] = indexed  # remember status
+                if name_plus_cpi in used_names:
+                    msg = 'parameter {} used twice in year_mods for year {}'
+                    raise ValueError(msg.format(name_plus_cpi, year))
+                else:
+                    used_names.add(name_plus_cpi)
+            else:
+                indexed = vals_indexed
             # set post-reform values of parameter with name
+            if name in used_names:
+                msg = 'parameter {} used twice in year_mods for year {}'
+                raise ValueError(msg.format(name, year))
+            else:
+                used_names.add(name)
             cval = getattr(self, name, None)
             if cval is None:
-                continue
+                msg = 'parameter {} in year_mods for year [] is unknown'
+                raise ValueError(msg.format(name, year))
             nval = self.expand_array(values,
-                                     inflate=cpi_inflated,
+                                     inflate=indexed,
                                      inflation_rates=inf_rates,
                                      num_years=num_years_to_expand)
-            cval[(self.current_year - self.start_year):] = nval
-        self.set_year(self._current_year)
+            cval[(year - self.start_year):] = nval
+        # handle unused parameter names, which all end in _cpi
+        for name, values in year_mods[year].items():
+            if name not in used_names:
+                if not name.endswith('_cpi'):
+                    msg = 'unused parameter name {} does not end in _cpi'
+                    raise ValueError(msg.format(name))
+                pname = name[:-4]  # root parameter name
+                if pname not in self._vals:
+                    msg = 'root parameter name {} not in values dictionary'
+                    raise ValueError(msg.format(pname))
+                pindexed = values  # set indexing status
+                self._vals[pname]['cpi_inflated'] = pindexed  # remember status
+                pvals = getattr(self, pname)
+                pvalues = [pvals[year - self.start_year]]
```

Comment:
@talumbau,  If you remove the outer brackets as you suggested and use
```
pvalues = pvals[year - self.start_year]
```
you get the following error:
```
...
# handle unused parameter names, which all end in _cpi
        for name, pindexed in year_mods[year].items():
            if name not in used_names:
                if not name.endswith('_cpi'):
                    msg = 'unused parameter name {} does not end in _cpi'
                    raise ValueError(msg.format(name))
                pname = name[:-4]  # root parameter name
                if pname not in self._vals:
                    msg = 'root parameter name {} not in values dictionary'
                    raise ValueError(msg.format(pname))
                self._vals[pname]['cpi_inflated'] = pindexed  # remember status
                pvals = getattr(self, pname)
                pvalues = pvals[year - self.start_year]
                cval = getattr(self, pname, None)
                if cval is None:
                    msg = 'parameter {} in year_mods for year [] is unknown'
                    raise ValueError(msg.format(pname, year))
                nval = self.expand_array(pvalues,
                                         inflate=pindexed,
                                         inflation_rates=inf_rates,
                                         num_years=num_years_to_expand)
>               cval[(year - self.start_year):] = nval
E               ValueError: could not broadcast input array from shape (8) into shape (8,6)

parameters_base.py:353: ValueError
```
Does the code make more sense now?